ComSemApp/student/views.py

```diff
@@ -236,3 +246,277 @@ def form_valid(self, form):
+            if past_correct_review:
+                time_since_view = (datetime.date.today() - past_correct_review.latest("date").date.date()).days
+                expression_z = (statistics.mean([x.response_time for x in past_correct_review]) - course_avg)/course_std
+            else:
+                time_since_view = (datetime.date.today() - e.last_submission).days
+                expression_z = 0
+
+            # AF - placeholder algorithm: 1/(number of initial attempts + days since last seen  + 
+            #      expression reaction time vs course reaction time z score + number of incorrect review attempts - number of correct review attempts)
+            e.practice_score = int(100/(max(0, attempt_factor + time_since_view + expression_z + past_incorrect_count - len(past_correct_review)) + 1))
+
+            if e.practice_score <= 33:
+                e.bar_style = "bg-danger"
+                e.border_style = "border-danger"
+            elif e.practice_score <= 66:
+                e.bar_style = "bg-warning"
+                e.border_style = "border-warning"
+            else:
+                e.bar_style = "bg-primary"
+                e.border_style = "border-success"
+        
+        return expression_qset
+
+    def get_context_data(self, **kwargs):
+        """Get worksheet data for a student in a given course
+        
+        Returns:
+            dict -- context for creating generate_reviewsheet.html
+        """
+        context = super(ReviewsheetGeneratorView, self).get_context_data(**kwargs)
+        worksheets = self.course.worksheets.filter(status=teacher_constants.WORKSHEET_STATUS_RELEASED)
+        course_response_times = [x.response_time for x in ReviewAttempt.objects.filter(student=self.student, course=self.course, correct=True)]
+
+        # TODO should this logic be in the worksheet model ?
+        for worksheet in worksheets:
+            last_submission = worksheet.last_submission(self.student)
+            complete_submission = worksheet.complete_submission(self.student) # vhl checks for complete worksheets
+            complete_submission_status = "complete" if complete_submission else "none"
+    
+            worksheet.last_submission_status = last_submission.status if last_submission else "none"
+            if complete_submission_status == 'complete': # vhl if there is a complete worksheet
+                worksheet.last_submission_status = 'complete'
+            if worksheet.last_submission_status == 'complete':
+                worksheet.expression_list = self.get_expressions(worksheet, course_response_times)
+
+        # Only allow students to review completed worksheets (must have an answer)        
+        context['worksheets'] =  [x for x in worksheets if x.last_submission_status == 'complete']
+
+
+        return context
+
+class ReviewsheetView(StudentCourseViewMixin, DetailView):
+    # SHOULD THIS BE INHERETING ^^? 
```

Comment:
I think inheriting from StudentCourseViewMixin is the way to go. If a user who isn't a student in the course specified in the url tries to access the page then they'll be redirected.
I'm not so sure that this or `ReviewsheetGeneratorView` should be a `DetailView`s though, as their purpose isn't to show the details of a Course. My vote would be a TemplateView for both but I'm ok with leaving it the way it is.